File: scripts/extract-fa-icons.mjs
#!/usr/bin/env node
/**
 * Extract Font Awesome icons used in FPbase codebase
 * Generates a TypeScript file with SVG path data for all icons
 */

import { writeFileSync } from 'fs';
import { join } from 'path';

// Icon mapping: kebab-case class names → camelCase FA exports
const iconMap = {
  // Solid icons
  'arrow-left': { pkg: 'solid', name: 'faArrowLeft' },
  'arrow-right': { pkg: 'solid', name: 'faArrowRight' },
  'arrows': { pkg: 'solid', name: 'faArrows' },
  'bars': { pkg: 'solid', name: 'faBars' },
  'book': { pkg: 'solid', name: 'faBook' },
  'camera': { pkg: 'solid', name: 'faCamera' },
  'chart-area': { pkg: 'solid', name: 'faChartArea' },
  'check': { pkg: 'solid', name: 'faCheck' },
  'check-circle': { pkg: 'solid', name: 'faCheckCircle' },
  'clock': { pkg: 'solid', name: 'faClock' },
  'code': { pkg: 'solid', name: 'faCode' },
  'cog': { pkg: 'solid', name: 'faCog' },
  'dot-circle-o': { pkg: 'regular', name: 'faDotCircle' }, // v4 → v5
  'download': { pkg: 'solid', name: 'faDownload' },
  'edit': { pkg: 'solid', name: 'faEdit' },
  'envelope': { pkg: 'solid', name: 'faEnvelope' },
  'exchange': { pkg: 'solid', name: 'faExchange' },
  'exchange-alt': { pkg: 'solid', name: 'faExchangeAlt' },
  'exclamation-circle': { pkg: 'solid', name: 'faExclamationCircle' },
  'exclamation-triangle': { pkg: 'solid', name: 'faExclamationTriangle' },
  'external-link-alt': { pkg: 'solid', name: 'faExternalLinkAlt' },
  'eye': { pkg: 'solid', name: 'faEye' },
  'fast-backward': { pkg: 'solid', name: 'faFastBackward' },
  'fast-forward': { pkg: 'solid', name: 'faFastForward' },
  'filter': { pkg: 'solid', name: 'faFilter' },
  'flag': { pkg: 'solid', name: 'faFlag' },
  'flash': { pkg: 'solid', name: 'faBolt' }, // v4 → v5 renamed
  'flip-horizontal': { pkg: 'solid', name: 'faArrowsLeftRight' }, // closest match
  'floppy-o': { pkg: 'solid', name: 'faFloppyDisk' }, // v4 → v5
  'heart': { pkg: 'solid', name: 'faHeart' },
  'heart-o': { pkg: 'regular', name: 'faHeart', exportName: 'farHeart' }, // outline heart for favorites
  'home': { pkg: 'solid', name: 'faHome' },
  'info': { pkg: 'solid', name: 'faInfo' },
  'info-circle': { pkg: 'solid', name: 'faInfoCircle' },
  'key': { pkg: 'solid', name: 'faKey' },
  'keyboard': { pkg: 'solid', name: 'faKeyboard' },
  'lightbulb': { pkg: 'solid', name: 'faLightbulb' },
  'link': { pkg: 'solid', name: 'faLink' },
  'list': { pkg: 'solid', name: 'faList' },
  'minus': { pkg: 'solid', name: 'faMinus' },
  'minus-circle': { pkg: 'solid', name: 'faMinusCircle' },
  'pause': { pkg: 'solid', name: 'faPause' },
  'play': { pkg: 'solid', name: 'faPlay' },
  'plus': { pkg: 'solid', name: 'faPlus' },
  'plus-circle': { pkg: 'solid', name: 'faPlusCircle' },
  'power-off': { pkg: 'solid', name: 'faPowerOff' },
  'question-circle': { pkg: 'solid', name: 'faQuestionCircle' },
  'quote-left': { pkg: 'solid', name: 'faQuoteLeft' },
  'search': { pkg: 'solid', name: 'faSearch' },
  'share': { pkg: 'solid', name: 'faShare' },
  'share-square': { pkg: 'solid', name: 'faShareSquare' },
  'spinner': { pkg: 'solid', name: 'faSpinner' },
  'square': { pkg: 'solid', name: 'faSquare' },
  'square-o': { pkg: 'regular', name: 'faSquare' }, // v4 → v5 regular
  'star': { pkg: 'solid', name: 'faStar' },
  'step-backward': { pkg: 'solid', name: 'faStepBackward' },
  'step-forward': { pkg: 'solid', name: 'faStepForward' },
  'sun': { pkg: 'solid', name: 'faSun' },
  'table': { pkg: 'solid', name: 'faTable' },
  'th': { pkg: 'solid', name: 'faTh' },
  'thumbtack': { pkg: 'solid', name: 'faThumbtack' },
  'times': { pkg: 'solid', name: 'faTimes' },
  'times-circle': { pkg: 'solid', name: 'faTimesCircle' },
  'trash': { pkg: 'solid', name: 'faTrash' },
  'trash-alt': { pkg: 'solid', name: 'faTrashAlt' },
  'undo': { pkg: 'solid', name: 'faUndo' },
  'upload': { pkg: 'solid', name: 'faUpload' },
  'user': { pkg: 'solid', name: 'faUser' },
  'wrench': { pkg: 'solid', name: 'faWrench' },

  // Brand icons
  'github': { pkg: 'brands', name: 'faGithub' },
  'google': { pkg: 'brands', name: 'faGoogle' },
  'orcid': { pkg: 'brands', name: 'faOrcid' },
  'x-twitter': { pkg: 'brands', name: 'faXTwitter' },
};

async function extractIcons() {
  const icons = [];
  const errors = [];

  for (const [kebabName, config] of Object.entries(iconMap)) {
    const { pkg, name, exportName } = config;
    try {
      const pkgName = `@fortawesome/free-${pkg}-svg-icons`;
      const module = await import(pkgName);
      const icon = module[name];

      if (!icon) {
        errors.push(`Icon ${name} not found in ${pkgName}`);
        continue;
      }

      const [width, height, , , svgPathData] = icon.icon;

      // Handle both single path and array of paths
      const path = Array.isArray(svgPathData) ? svgPathData.join(' ') : svgPathData;

      // Generate consistent prefix based on package: fas/far/fab
      const prefix = pkg === 'solid' ? 'fas' : pkg === 'regular' ? 'far' : 'fab';
      const consistentName = exportName || name.replace(/^fa/, prefix);

      icons.push({
        kebabName,
        exportName: consistentName,
        width,
        height,
        path,
      });
    } catch (err) {
      errors.push(`Error importing ${name}: ${err.message}`);
    }
  }

  if (errors.length > 0) {
    console.error('Errors encountered:');
    errors.forEach(err => console.error(`  - ${err}`));
  }

  return icons;
}

function generateTypeScript(icons) {
  const header = `/**
 * Static FontAwesome icon data
 * Extracted from @fortawesome packages to eliminate runtime dependency
 *
 * Icon data format: { width, height, path }
 * - width/height: SVG viewBox dimensions
 * - path: SVG path data string
 *
 * Generated by scripts/extract-fa-icons.mjs
 */

export interface IconData {
  width: number
  height: number
  path: string
}
`;

  const iconExports = icons
    .sort((a, b) => a.exportName.localeCompare(b.exportName))
    .map(({ exportName, width, height, path }) => {
      // Escape quotes in path data
      const escapedPath = path.replace(/\\/g, '\\\\').replace(/"/g, '\\"');

      return `export const ${exportName}: IconData = {
  width: ${width},
  height: ${height},
  path: "${escapedPath}",
}`;
    })
    .join('\n\n');

  return header + '\n' + iconExports + '\n';
}

function generatePython(icons) {
  const header = `"""
Static FontAwesome icon data
Extracted from @fortawesome packages to eliminate runtime dependency

Icon data format: dict with 'width', 'height', 'path' keys
- width/height: SVG viewBox dimensions
- path: SVG path data string

Generated by scripts/extract-fa-icons.mjs
"""

from __future__ import annotations

from typing import TypedDict


class IconData(TypedDict):
    """SVG icon data."""

    width: int
    height: int
    path: str


`;

  // Create dict entries
  const iconDict = icons
    .sort((a, b) => a.exportName.localeCompare(b.exportName))
    .map(({ exportName, kebabName, width, height, path }) => {
      // Escape quotes and backslashes for Python
      const escapedPath = path.replace(/\\/g, '\\\\').replace(/"/g, '\\"');

      return `    "${exportName}": {
        "width": ${width},
        "height": ${height},
        "path": "${escapedPath}",
    }`;
    })
    .join(',\n');

  // Also create a kebab-case alias mapping for template tag usage
  const aliasMapping = icons
    .sort((a, b) => a.kebabName.localeCompare(b.kebabName))
    .map(({ exportName, kebabName }) => `    "${kebabName}": "${exportName}"`)
    .join(',\n');

  return `${header}
ICONS: dict[str, IconData] = {
${iconDict}
}

# Kebab-case aliases for template tag convenience
# Maps 'trash' → 'fasTrash', 'square-o' → 'farSquare', etc.
ICON_ALIASES: dict[str, str] = {
${aliasMapping}
}


def get_icon(name: str) -> IconData | None:
    """
    Get icon data by name.

    Accepts both camelCase exports (fasTrash) and kebab-case aliases (trash).
    """
    # Try direct lookup first
    if name in ICONS:
        return ICONS[name]

    # Try alias lookup
    if name in ICON_ALIASES:
        return ICONS[ICON_ALIASES[name]]

    return None
`;
}

async function main() {
  console.log('Extracting Font Awesome icons...');
  const icons = await extractIcons();
  console.log(`✓ Extracted ${icons.length} icons`);

  // Generate TypeScript for frontend
  const tsContent = generateTypeScript(icons);
  const tsPath = join(process.cwd(), 'frontend/src/icons/fa-icons.ts');
  writeFileSync(tsPath, tsContent, 'utf8');
  console.log(`✓ Written TypeScript to ${tsPath}`);

  // Generate Python for Django templates
  const pyContent = generatePython(icons);
  const pyPath = join(process.cwd(), 'backend/fpbase/icons.py');
  writeFileSync(pyPath, pyContent, 'utf8');
  console.log(`✓ Written Python to ${pyPath}`);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/extract_fa_icons.py
# extract_fontawesome_svgs.py
import argparse
import re
import shutil
import tempfile
import zipfile
from pathlib import Path

import requests

VERSION = "7.1.0"
URL = f"https://github.com/FortAwesome/Font-Awesome/releases/download/{VERSION}/fontawesome-free-{VERSION}-web.zip"

ICON_MAP = {
    # UI & Navigation
    "info": ("solid", "info-circle"),
    "info-i": ("solid", "info"),
    "warning": ("solid", "exclamation-circle"),
    "alert": ("solid", "exclamation-triangle"),
    "help": ("solid", "info-circle"),
    "question": ("solid", "question-circle"),
    "close": ("solid", "times"),
    "remove": ("solid", "times-circle"),
    "menu": ("solid", "list"),
    "grid": ("solid", "th"),
    "search": ("solid", "search"),
    "filter": ("solid", "filter"),
    "view": ("solid", "eye"),
    "settings": ("solid", "cog"),
    "edit": ("solid", "edit"),
    "delete": ("solid", "trash-alt"),
    "trash": ("solid", "trash"),
    "undo": ("solid", "undo"),
    "check": ("solid", "check"),
    "success": ("solid", "check-circle"),
    "selected": ("regular", "check-square"),
    "unselected": ("regular", "square"),
    "heart": ("solid", "heart"),
    "heart-outline": ("regular", "heart"),
    "add": ("solid", "plus"),
    "add-item": ("solid", "plus-circle"),
    "download": ("solid", "download"),
    "upload": ("solid", "upload"),
    "share": ("solid", "share"),
    "share-square": ("solid", "share-square"),
    "link": ("solid", "link"),
    "external-link": ("solid", "external-link-alt"),
    "exchange": ("solid", "exchange-alt"),
    "book": ("solid", "book"),
    "collection": ("solid", "book"),
    "quote": ("solid", "quote-left"),
    "photo": ("solid", "camera"),
    "chart": ("solid", "chart-area"),
    "table": ("solid", "table"),
    "flag": ("solid", "flag"),
    "flag-outline": ("regular", "flag"),
    "clock": ("solid", "clock"),
    "spinner": ("solid", "spinner"),
    "lightbulb": ("solid", "lightbulb"),
    "sun": ("solid", "sun"),
    "email": ("solid", "envelope"),
    "wrench": ("solid", "wrench"),
    "keyboard": ("regular", "keyboard"),
    "google": ("brands", "google"),
    "twitter": ("brands", "twitter"),
    "orcid": ("brands", "orcid"),
}


XML_COMMENT_RE = re.compile(r"<!--.*?-->", re.DOTALL)


def _copy_icon(src: Path, dest: Path) -> None:
    content = src.read_text(encoding="utf-8")
    content = XML_COMMENT_RE.sub("", content)
    dest.write_text(content, encoding="utf-8")


def extract_svgs(
    outdir: str | Path = "svgs", extract_dir_name: str = "fa-extract", keep_existing: bool = False
) -> None:
    """Download Font Awesome zip, extract into a named temp subfolder, and copy selected SVGs.

    Args:
        outdir: destination folder (created if missing) where the selected SVGs will be copied.
        extract_dir_name: the subfolder name under the temporary directory where the zip will be extracted.
    """
    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir_path = Path(tmpdir)
        extract_path = tmpdir_path / extract_dir_name
        extract_path.mkdir(parents=True, exist_ok=True)
        zip_path = tmpdir_path / "fa.zip"

        print(f"Downloading {URL}...")
        with requests.get(URL, stream=True) as r:
            r.raise_for_status()
            with open(zip_path, "wb") as f:
                for chunk in r.iter_content(8192):
                    f.write(chunk)

        print("Extracting...")
        with zipfile.ZipFile(zip_path) as z:
            # Extract into a named subfolder so caller can control where files land
            z.extractall(extract_path)

        # Find top-level extracted Font-Awesome-* folder
        try:
            svg_path = next(extract_path.glob("fontawesome-*/svgs/"))
        except StopIteration:
            raise RuntimeError(
                "Could not find extracted Font Awesome folder. Check the structure of the zip file."
            ) from None

        outdir = Path(outdir)
        if outdir.exists() and not keep_existing:
            print("Removing existing SVGs...")
            shutil.rmtree(outdir, ignore_errors=True)
        outdir.mkdir(parents=True, exist_ok=True)

        for key, (style, glyph) in ICON_MAP.items():
            src = svg_path / style / f"{glyph}.svg"
            if src.exists():
                _copy_icon(src, outdir / f"{key}.svg")
                print(f"Copied {style}/{src.name} -> {key}.svg")
            else:
                raise ValueError(f"Missing: {glyph} ({style})")

    print("\n✅ Done")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Download Font Awesome and extract selected SVGs.")
    parser.add_argument(
        "--outdir", "-o", default="backend/fpbase/static/icons", help="Destination folder for copied SVGs"
    )
    parser.add_argument(
        "--extract-dir-name",
        "-e",
        default="fa-extract",
        help="Name of temporary subfolder to extract the zip into",
    )
    parser.add_argument(
        "--keep-existing",
        action="store_true",
        help="If set, existing SVGs in the output directory will not be overwritten",
    )
    args = parser.parse_args()

    extract_svgs(outdir=args.outdir, extract_dir_name=args.extract_dir_name, keep_existing=args.keep_existing)
